File: arc-agi-runner/src/worldModel/clickFeatures.ts
/* [arc-agi-runner/worldModel/clickFeatures] BL.21560 -- vector de features POR CELDA para decidir
   DONDE clickear (ACTION6), y su combinacion lineal con los pesos de `clickPriors.ts`.

   EL PROBLEMA, medido sobre el corpus real (`arcReplayFrames`, corrida ft09-0d8bbf25):
   346 clicks, 32 productivos (9,2%). La heuristica previa (`pickClickTarget`) elegia UNIFORMEMENTE
   entre el decil superior de celdas "borde de color" -- unos 410 candidatos de 4096, de los cuales
   solo ~36 son las esquinas de las fichas del tablero jugable. 36/410 = 8,8%: la tasa medida no era
   mala suerte, era exactamente lo que predice tirar al azar dentro de ese conjunto.

   POR QUE ESTAS FEATURES Y NO OTRAS. Todas salen de helpers que YA existen (`grid.ts` +
   `findComponents` de `primitiveOps.ts`): son la misma segmentacion que ve el modelo de mundo, no
   una vision paralela. Y ninguna mira coordenadas absolutas ni identificadores de partida -- un
   peso aprendido en un juego tiene que poder servir en otro, que es la unica razon para transportar
   pesos. La feature que mas separa en el corpus real es `componenteRodeadaDeFondo`: el mismo dibujo
   (una ficha 6x6 de color 9) aparece DOS veces en pantalla -- como panel decorativo rodeado por el
   fondo, y como ficha jugable rodeada por el marco del tablero. Clickear la primera no hace nada
   (62 clicks muertos medidos); clickear la segunda siempre funciona (32 de 32). Nada del color ni
   del tamano distingue una de la otra: lo hace el VECINDARIO de su componente.

   Los pesos NO se escriben a mano: los ajusta `arc-agi3-kaggle-agent/scripts/fit_click_priors.py`
   por regresion logistica contra el corpus real y los emite a `clickPriors.ts` (este puerto) y a
   `arc_agent/priors.py` (el puerto Python). */

import {
  detectBackgroundColor,
  foregroundBoundingBox,
  type BoundingBox,
  type Grid,
} from './grid';
import { findComponents } from './primitiveOps';

/** ORDEN CANONICO de las features -- es un CONTRATO con `clickPriors.ts` y con el puerto Python
 *  (`arc_agent/click_targeting.py`): los pesos son un array posicional, asi que reordenar esta
 *  lista sin regenerar los priors invierte silenciosamente el significado de cada peso. El primer
 *  elemento es el sesgo (siempre 1). */
export const CLICK_FEATURE_NAMES = [
  'sesgo',
  'bordeDeColor',
  'tamanoComponente',
  'esBordeDeComponente',
  'rarezaDeColor',
  'esColorDeFondo',
  'distanciaAlBboxDeForeground',
  'componenteRodeadaDeFondo',
  'enRegionQueCambio',
] as const;

export const CLICK_FEATURE_COUNT = CLICK_FEATURE_NAMES.length;

/** Normalizador de `tamanoComponente`. 256 celdas = 1/16 de un frame 64x64: por encima de eso la
 *  "componente" ya es una region de fondo o un panel entero, no un objeto clickeable. Satura en 1
 *  para que un fondo gigante no domine el producto punto por magnitud. */
const TAMANO_COMPONENTE_SATURACION = 256;

/** Radio del parche que se guarda como plantilla al detectar un click con efecto (3x3 = radio 1).
 *  Fuente unica: lo usan `clickMemory.ts` y su espejo Python. */
export const RADIO_PARCHE = 1;

/** Valor de una celda del parche que cae FUERA de la grilla. -1 no colisiona con ningun color ARC
 *  (0-15), asi que dos parches de borde solo matchean si el borde cae en el mismo lugar. */
export const PARCHE_FUERA_DE_GRILLA = -1;

export interface ClickFeatureBoardOptions {
  /** Rectangulo que cambio en la ULTIMA transicion observada, o `null` si no hubo/no se sabe.
   *  Alimenta la feature `enRegionQueCambio`: lo que se acaba de mover suele ser lo que responde. */
  regionCambiada?: BoundingBox | null;
}

/** Features ya calculadas para TODAS las celdas de una grilla. Se construye UNA vez por frame:
 *  segmentar en componentes es O(celdas) y hacerlo por candidato seria cuadratico. */
export interface ClickFeatureBoard {
  readonly ancho: number;
  readonly alto: number;
  readonly colorDeFondo: number;
  /** Vector de features de (x,y), en el orden de `CLICK_FEATURE_NAMES`. */
  features(x: number, y: number): number[];
  /** Tamano de la componente 4-conexa de (x,y); 0 si la celda es del color de fondo. */
  tamanoDeComponente(x: number, y: number): number;
}

function bboxContiene(bbox: BoundingBox, x: number, y: number): boolean {
  return x >= bbox.minX && x <= bbox.maxX && y >= bbox.minY && y <= bbox.maxY;
}

/** Distancia Chebyshev de (x,y) al rectangulo, 0 si esta adentro. */
function distanciaAlBBox(bbox: BoundingBox, x: number, y: number): number {
  const dx = Math.max(bbox.minX - x, 0, x - bbox.maxX);
  const dy = Math.max(bbox.minY - y, 0, y - bbox.maxY);
  return Math.max(dx, dy);
}

export function construirTableroDeFeatures(
  grid: Grid,
  opts: ClickFeatureBoardOptions = {},
): ClickFeatureBoard {
  const alto = grid.length;
  const ancho = alto > 0 ? (grid[0]?.length ?? 0) : 0;
  const total = ancho * alto;
  const colorDeFondo = detectBackgroundColor(grid);
  const fgBBox = foregroundBoundingBox(grid, colorDeFondo);
  const regionCambiada = opts.regionCambiada ?? null;

  /* Conteo global de colores -- alimenta `rarezaDeColor`. */
  const conteoDeColor = new Map<number, number>();
  for (let y = 0; y < alto; y++) {
    const fila = grid[y];
    for (let x = 0; x < ancho; x++) {
      const c = fila[x];
      conteoDeColor.set(c, (conteoDeColor.get(c) ?? 0) + 1);
    }
  }

  /* Segmentacion en componentes 4-conexas -- la MISMA que usa el DSL (objectExtract). */
  const etiqueta = new Int32Array(total).fill(-1);
  const componentes = findComponents(grid, colorDeFondo);
  const tamanos = new Int32Array(componentes.length);
  for (let i = 0; i < componentes.length; i++) {
    tamanos[i] = componentes[i].length;
    for (const [cx, cy] of componentes[i]) etiqueta[cy * ancho + cx] = i;
  }

  /* Fraccion del CONTORNO de cada componente que toca el color de fondo. Es lo que separa un panel
     decorativo (flotando sobre el fondo) de una ficha dentro de un tablero (rodeada por el marco). */
  const rodeadaDeFondo = new Float64Array(componentes.length);
  for (let i = 0; i < componentes.length; i++) {
    let contorno = 0;
    let contornoDeFondo = 0;
    for (const [cx, cy] of componentes[i]) {
      const vecinos: Array<[number, number]> = [
        [cx - 1, cy],
        [cx + 1, cy],
        [cx, cy - 1],
        [cx, cy + 1],
      ];
      for (const [vx, vy] of vecinos) {
        const dentro = vx >= 0 && vy >= 0 && vx < ancho && vy < alto;
        if (dentro && etiqueta[vy * ancho + vx] === i) continue;
        contorno++;
        if (dentro && grid[vy][vx] === colorDeFondo) contornoDeFondo++;
      }
    }
    rodeadaDeFondo[i] = contorno > 0 ? contornoDeFondo / contorno : 0;
  }

  const maxDistancia = Math.max(ancho, alto, 1);

  return {
    ancho,
    alto,
    colorDeFondo,
    tamanoDeComponente(x: number, y: number): number {
      if (x < 0 || y < 0 || x >= ancho || y >= alto) return 0;
      const id = etiqueta[y * ancho + x];
      return id < 0 ? 0 : tamanos[id];
    },
    features(x: number, y: number): number[] {
      if (x < 0 || y < 0 || x >= ancho || y >= alto) {
        return new Array<number>(CLICK_FEATURE_COUNT).fill(0);
      }
      const valor = grid[y][x];
      const id = etiqueta[y * ancho + x];

      let bordeDeColor = 0;
      if (x > 0 && grid[y][x - 1] !== valor) bordeDeColor++;
      if (x < ancho - 1 && grid[y][x + 1] !== valor) bordeDeColor++;
      if (y > 0 && grid[y - 1][x] !== valor) bordeDeColor++;
      if (y < alto - 1 && grid[y + 1][x] !== valor) bordeDeColor++;

      const tamano = id < 0 ? 0 : tamanos[id];
      /* Borde de la COMPONENTE, no de la grilla: una celda cuya componente se corta contra el
         limite del frame tambien cuenta (el vecino inexistente no pertenece a la componente). */
      const esBorde =
        id >= 0 &&
        (x === 0 ||
          y === 0 ||
          x === ancho - 1 ||
          y === alto - 1 ||
          etiqueta[y * ancho + (x - 1)] !== id ||
          etiqueta[y * ancho + (x + 1)] !== id ||
          etiqueta[(y - 1) * ancho + x] !== id ||
          etiqueta[(y + 1) * ancho + x] !== id);

      const conteo = conteoDeColor.get(valor) ?? 0;
      const rareza = total > 0 ? 1 - conteo / total : 0;

      return [
        1,
        bordeDeColor / 4,
        Math.min(1, tamano / TAMANO_COMPONENTE_SATURACION),
        esBorde ? 1 : 0,
        rareza,
        valor === colorDeFondo ? 1 : 0,
        fgBBox === null ? 1 : distanciaAlBBox(fgBBox, x, y) / maxDistancia,
        id < 0 ? 0 : rodeadaDeFondo[id],
        regionCambiada !== null && bboxContiene(regionCambiada, x, y) ? 1 : 0,
      ];
    },
  };
}

/** Producto punto features x pesos (logit). Pesos de largo distinto al vector se recortan/rellenan
 *  con 0: un `priors.ts` regenerado con una feature de mas nunca debe hacer crashear una partida,
 *  degrada a ignorar lo que no entiende. */
export function puntuarCelda(features: number[], pesos: readonly number[]): number {
  let acumulado = 0;
  const n = Math.min(features.length, pesos.length);
  for (let i = 0; i < n; i++) acumulado += features[i] * pesos[i];
  return acumulado;
}

/** Probabilidad logistica -- solo para reportar/umbralar; el ranking usa el logit directo. */
export function sigmoide(logit: number): number {
  if (logit >= 0) return 1 / (1 + Math.exp(-logit));
  const e = Math.exp(logit);
  return e / (1 + e);
}

/** Parche cuadrado de lado `2*RADIO_PARCHE+1` centrado en (x,y), aplanado en orden row-major.
 *  Las celdas fuera de la grilla valen `PARCHE_FUERA_DE_GRILLA`. */
export function extraerParche(grid: Grid, x: number, y: number): number[] {
  const alto = grid.length;
  const ancho = alto > 0 ? (grid[0]?.length ?? 0) : 0;
  const parche: number[] = [];
  for (let dy = -RADIO_PARCHE; dy <= RADIO_PARCHE; dy++) {
    for (let dx = -RADIO_PARCHE; dx <= RADIO_PARCHE; dx++) {
      const px = x + dx;
      const py = y + dy;
      const dentro = px >= 0 && py >= 0 && px < ancho && py < alto;
      parche.push(dentro ? grid[py][px] : PARCHE_FUERA_DE_GRILLA);
    }
  }
  return parche;
}

/** Similitud entre dos parches: fraccion de celdas iguales, en [0,1]. Largos distintos (imposible
 *  hoy, pero barato de blindar) comparan solo el prefijo comun. */
export function similitudDeParche(a: readonly number[], b: readonly number[]): number {
  const n = Math.min(a.length, b.length);
  if (n === 0) return 0;
  let iguales = 0;
  for (let i = 0; i < n; i++) if (a[i] === b[i]) iguales++;
  return iguales / n;
}

/** Rectangulo que cambio entre dos grillas, o `null` si no cambio nada (o si no son comparables).
 *  Alimenta la feature `enRegionQueCambio`. */
export function regionQueCambio(pre: Grid | null, post: Grid | null): BoundingBox | null {
  if (pre === null || post === null) return null;
  let minX = Infinity;
  let minY = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;
  const alto = Math.min(pre.length, post.length);
  for (let y = 0; y < alto; y++) {
    const ancho = Math.min(pre[y]?.length ?? 0, post[y]?.length ?? 0);
    for (let x = 0; x < ancho; x++) {
      if (pre[y][x] === post[y][x]) continue;
      if (x < minX) minX = x;
      if (x > maxX) maxX = x;
      if (y < minY) minY = y;
      if (y > maxY) maxY = y;
    }
  }
  if (maxX < minX) return null;
  return { minX, minY, maxX, maxY };
}

File: arc-agi3-kaggle-agent/arc_agent/click_features.py
"""[arc-agi3-kaggle-agent/click_features] BL.21560 -- vector de features POR CELDA para decidir donde
clickear (ACTION6) y su combinacion lineal con los pesos de priors.py. Espejo de
arc-agi-runner/src/worldModel/clickFeatures.ts.

POR QUE ESTAS FEATURES. Todas salen de helpers que YA existen (`grid.py` + `_find_components`): es la
misma segmentacion que ve el modelo de mundo, no una vision paralela. Y ninguna mira coordenadas
absolutas ni identificadores de partida -- un peso aprendido en un juego tiene que poder servir en
otro, que es la unica razon para transportar pesos.

La feature que mas separa en el corpus real es `componenteRodeadaDeFondo`: el mismo dibujo (una ficha
6x6 de color 9) aparece DOS veces en pantalla de ft09 -- como panel decorativo rodeado por el fondo, y
como ficha jugable rodeada por el marco del tablero. Clickear la primera no hace nada (62 clicks
muertos medidos); clickear la segunda funciono 32 de 32 veces. Nada del color ni del tamano las
distingue: lo hace el VECINDARIO de su componente. OJO con generalizar de mas -- en lp85-305b61c3 la
relacion es la inversa, y por eso el peso de esta feature esta regularizado y la evidencia del
episodio (plantillas de click_targeting.py) pesa mas que el prior.
"""
from __future__ import annotations

import math
from typing import Sequence

# Import a UN solo nivel (`.world_model`): el builder del notebook desmonta los imports relativos
# con el regex `^from \.\w* import .+$`, que no cubre un segundo punto.
from .world_model import Grid, detect_background_color, foreground_bounding_box
from .world_model import _find_components


CLICK_FEATURE_NAMES: tuple[str, ...] = (
    "sesgo",
    "bordeDeColor",
    "tamanoComponente",
    "esBordeDeComponente",
    "rarezaDeColor",
    "esColorDeFondo",
    "distanciaAlBboxDeForeground",
    "componenteRodeadaDeFondo",
    "enRegionQueCambio",
)

CLICK_FEATURE_COUNT = len(CLICK_FEATURE_NAMES)

#: Normalizador de `tamanoComponente`. 256 celdas = 1/16 de un frame 64x64: por encima de eso la
#: "componente" ya es una region de fondo o un panel entero, no un objeto clickeable.
TAMANO_COMPONENTE_SATURACION = 256


def _bbox_contiene(bbox: tuple[int, int, int, int], x: int, y: int) -> bool:
    min_x, min_y, max_x, max_y = bbox
    return min_x <= x <= max_x and min_y <= y <= max_y


def _distancia_al_bbox(bbox: tuple[int, int, int, int], x: int, y: int) -> int:
    """Distancia Chebyshev al rectangulo, 0 si (x,y) esta adentro."""
    min_x, min_y, max_x, max_y = bbox
    dx = max(min_x - x, 0, x - max_x)
    dy = max(min_y - y, 0, y - max_y)
    return max(dx, dy)


class ClickFeatureBoard:
    """Features de TODAS las celdas de una grilla, calculadas una sola vez por frame: segmentar en
    componentes es O(celdas) y rehacerlo por candidato seria cuadratico."""

    __slots__ = (
        "_grid",
        "ancho",
        "alto",
        "color_de_fondo",
        "_etiqueta",
        "_tamanos",
        "_rodeada_de_fondo",
        "_conteo_de_color",
        "_fg_bbox",
        "_region_cambiada",
        "_total",
    )

    def __init__(
        self, grid: Grid, region_cambiada: tuple[int, int, int, int] | None = None
    ) -> None:
        self._grid = grid
        self.alto = len(grid)
        self.ancho = len(grid[0]) if self.alto > 0 else 0
        self._total = self.ancho * self.alto
        self.color_de_fondo = detect_background_color(grid)
        caja = foreground_bounding_box(grid, self.color_de_fondo)
        self._fg_bbox = (
            None if caja is None else (caja.min_x, caja.min_y, caja.max_x, caja.max_y)
        )
        self._region_cambiada = region_cambiada

        conteo: dict[int, int] = {}
        for fila in grid:
            for celda in fila:
                conteo[celda] = conteo.get(celda, 0) + 1
        self._conteo_de_color = conteo

        etiqueta = [-1] * self._total
        componentes = _find_components(grid, self.color_de_fondo)
        tamanos = [len(c) for c in componentes]
        for i, celdas in enumerate(componentes):
            for cx, cy in celdas:
                etiqueta[cy * self.ancho + cx] = i
        self._etiqueta = etiqueta
        self._tamanos = tamanos

        # Fraccion del CONTORNO de cada componente que toca el color de fondo. Es lo que separa un
        # panel decorativo (flotando sobre el fondo) de una ficha dentro de un tablero (rodeada por
        # el marco): en ft09 el MISMO dibujo aparece de las dos formas y solo la segunda responde al
        # click -- 32 de 32 productivos contra 62 muertos sobre el dibujo gemelo.
        rodeada = [0.0] * len(componentes)
        for i, celdas in enumerate(componentes):
            contorno = 0
            contorno_de_fondo = 0
            for cx, cy in celdas:
                for vx, vy in ((cx - 1, cy), (cx + 1, cy), (cx, cy - 1), (cx, cy + 1)):
                    dentro = 0 <= vx < self.ancho and 0 <= vy < self.alto
                    if dentro and etiqueta[vy * self.ancho + vx] == i:
                        continue
                    contorno += 1
                    if dentro and grid[vy][vx] == self.color_de_fondo:
                        contorno_de_fondo += 1
            rodeada[i] = contorno_de_fondo / contorno if contorno > 0 else 0.0
        self._rodeada_de_fondo = rodeada

    def tamano_de_componente(self, x: int, y: int) -> int:
        if not (0 <= x < self.ancho and 0 <= y < self.alto):
            return 0
        ident = self._etiqueta[y * self.ancho + x]
        return 0 if ident < 0 else self._tamanos[ident]

    def features(self, x: int, y: int) -> list[float]:
        if not (0 <= x < self.ancho and 0 <= y < self.alto):
            return [0.0] * CLICK_FEATURE_COUNT
        grid = self._grid
        valor = grid[y][x]
        ident = self._etiqueta[y * self.ancho + x]

        borde_de_color = 0
        if x > 0 and grid[y][x - 1] != valor:
            borde_de_color += 1
        if x + 1 < self.ancho and grid[y][x + 1] != valor:
            borde_de_color += 1
        if y > 0 and grid[y - 1][x] != valor:
            borde_de_color += 1
        if y + 1 < self.alto and grid[y + 1][x] != valor:
            borde_de_color += 1

        tamano = 0 if ident < 0 else self._tamanos[ident]
        # Borde de la COMPONENTE, no de la grilla: una celda cuya componente se corta contra el
        # limite del frame tambien cuenta (el vecino inexistente no pertenece a la componente).
        es_borde = ident >= 0 and (
            x == 0
            or y == 0
            or x == self.ancho - 1
            or y == self.alto - 1
            or self._etiqueta[y * self.ancho + (x - 1)] != ident
            or self._etiqueta[y * self.ancho + (x + 1)] != ident
            or self._etiqueta[(y - 1) * self.ancho + x] != ident
            or self._etiqueta[(y + 1) * self.ancho + x] != ident
        )

        conteo = self._conteo_de_color.get(valor, 0)
        rareza = 1 - conteo / self._total if self._total > 0 else 0.0
        max_distancia = max(self.ancho, self.alto, 1)

        return [
            1.0,
            borde_de_color / 4,
            min(1.0, tamano / TAMANO_COMPONENTE_SATURACION),
            1.0 if es_borde else 0.0,
            rareza,
            1.0 if valor == self.color_de_fondo else 0.0,
            1.0
            if self._fg_bbox is None
            else _distancia_al_bbox(self._fg_bbox, x, y) / max_distancia,
            0.0 if ident < 0 else self._rodeada_de_fondo[ident],
            1.0
            if self._region_cambiada is not None
            and _bbox_contiene(self._region_cambiada, x, y)
            else 0.0,
        ]


def puntuar_celda(features: Sequence[float], pesos: Sequence[float]) -> float:
    """Producto punto features x pesos (logit). Largos distintos se recortan al minimo comun: unos
    priors regenerados con una feature de mas nunca deben tumbar una partida."""
    return sum(f * p for f, p in zip(features, pesos))


def sigmoide(logit: float) -> float:
    """Probabilidad logistica -- solo para reportar/umbralar; el ranking usa el logit directo.
    Formulada por rama para no desbordar `exp` con logits grandes."""
    if logit >= 0:
        return 1 / (1 + math.exp(-logit))
    e = math.exp(logit)
    return e / (1 + e)
